### PR TITLE
dialects (wasm): Add attribute-based module sections

### PR DESCRIPTION
**Pull request by Moxinilian**
Adds the module sections that are attribute-based. Region-based sections are still missing.

### CODE REVIEW

File: tests/dialects/wasm/test_wasm.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+def test_partial_module():
+    module = WasmModule(
+        mems=[WasmLimits(0, None)],
+        imports=[WasmImport("test", WasmImportDescFunc(WasmFuncType((i32,), (i32,))))],
+    )
+    output = StringIO()
+    Printer(stream=output).print_op(module)
+
+    EXPECTED = """wasm.module(
```
Comment (superlopuh) [inline tests/dialects/wasm/test_wasm.py:35]: supernit:
```suggestion
    EXPECTED = """\
wasm.module(
```

File: xdsl/dialects/wasm/attrs.py [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+    elements: ParameterDef[WasmRefType]
+    limits: ParameterDef[WasmLimits]
+
+    def __init__(self, elem: WasmRefType, limits: WasmLimits):
+        super().__init__((elem, limits))
+
+
+class WasmMutEnum(StrEnum):
+    Const = auto()
+    Var = auto()
```
Comment (superlopuh) [inline xdsl/dialects/wasm/attrs.py:187]:
```suggestion
    CONST = auto()
    VAR = auto()
```

File: xdsl/dialects/wasm/ops.py [hunk 3]
```diff
@@ -73,10 +168,30 @@ def encode(self, ctx: WasmBinaryEncodingContext, io: BinaryIO) -> None:
...
 
+        # FIXME: implement encoder for module attributes
+        if (
+            len(self.tables.data)
+            + len(self.mems.data)
+            + len(self.exports)
+            + len(self.imports)
+            != 0
+        ):
+            raise NotImplementedError()
```
Comment (superlopuh) [inline xdsl/dialects/wasm/ops.py:179]: Could you please add a diagnostic filecheck test?
```suggestion
            raise DiagnosticEception("Module attributes printing not implemented yet")
```
SImilarly for the notimplemented below

File: tests/dialects/wasm/test_wasm.py
Comment (superlopuh) [inline tests/dialects/wasm/test_wasm.py:1]: what's the benefit of doing this in python vs filecheck? Or is it just to test the Python API?

File: tests/filecheck/dialects/wasm/ops.mlir [hunk 4]
```diff
@@ -5,14 +5,39 @@
...
+  exports [#wasm.export<"test_ex", #wasm.export_desc_mem<0 : i32>>]
+)
+
+// CHECK-NEXT: wasm.module(
+// CHECK-NEXT:    tables [#wasm.table<#wasm<reftype funcref>, #wasm.limits<0 : i32, 8 : i32>>]
+// CHECK-NEXT:    mems [#wasm.limits<0 : i32, unit>, #wasm.limits<0 : i32, 4096 : i32>]
+// CHECK-NEXT:    imports [#wasm.import<"test_im", #wasm.import_desc_func<#wasm.functype<(i32) -> (i32)>>>]
+// CHECK-NEXT:    exports [#wasm.export<"test_ex", #wasm.export_desc_mem<0 : i32>>]
+// CHECK-NEXT:  )
+// CHECK-GENERIC-NEXT: "wasm.module"() <{"tables" = [#wasm.table<#wasm<reftype funcref>, #wasm.limits<0 : i32, 8 : i32>>], "mems" = [#wasm.limits<0 : i32, unit>, #wasm.limits<0 : i32, 4096 : i32>], "imports" = [#wasm.import<"test_im", #wasm.import_desc_func<#wasm.functype<(i32) -> (i32)>>>], "exports" = [#wasm.export<"test_ex", #wasm.export_desc_mem<0 : i32>>]}> : () -> ()
```
Comment (superlopuh) [inline tests/filecheck/dialects/wasm/ops.mlir:36]: I've been adding the check lines below the input, is the convention to put them above instead?

File: xdsl/dialects/wasm/attrs.py [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+    FuncRef = "funcref"
+    """
+    wasm> The type funcref denotes the infinite union of all references to
+    functions, regardless of their function types.
+    """
+
+    ExternRef = "externref"
+    """
+    wasm> The type externref denotes the infinite union of all references to
+    objects owned. by the embedder and that can be passed into WebAssembly
```
Comment (superlopuh) [inline xdsl/dialects/wasm/attrs.py:72]:
```suggestion
    objects owned by the embedder and that can be passed into WebAssembly
```

File: tests/dialects/wasm/test_wasm.py
Comment (Moxinilian) [reply tests/dialects/wasm/test_wasm.py:1]: It is to test the Python API

File: xdsl/dialects/wasm/attrs.py (hunk 2 above)
Comment (Moxinilian) [reply xdsl/dialects/wasm/attrs.py:187]: Do you mean that this should be caps because they are constants?

File: xdsl/dialects/wasm/ops.py (hunk 3 above)
Comment (Moxinilian) [reply xdsl/dialects/wasm/ops.py:179]: What is a diagnostic filecheck test?

File: tests/filecheck/dialects/wasm/ops.mlir (hunk 4 above)
Comment (Moxinilian) [reply tests/filecheck/dialects/wasm/ops.mlir:36]: I have always seen them above. I had not paid attention to them being below in your file. I'd recommend putting them above.

File: xdsl/dialects/wasm/ops.py (hunk 3 above)
Comment (superlopuh) [reply xdsl/dialects/wasm/ops.py:179]: We have a few around, it's a way to test that a certain error is raised
```
// RUN: xdsl-opt %s --verify-diagnostics | filecheck %s
```

File: tests/dialects/wasm/test_wasm.py
Comment (superlopuh) [reply tests/dialects/wasm/test_wasm.py:1]: I'm not 100% sure we need it but I'm happy to leave it in

File: xdsl/dialects/wasm/attrs.py (hunk 2 above)
Comment (superlopuh) [reply xdsl/dialects/wasm/attrs.py:187]: Yep, this is the Python convention https://docs.python.org/3/library/enum.html

**Comment by superlopuh** [comment]
Let's close this?

**Pull request closed without merging**